pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a transform that computes the frequency spectrum of a quantum
+circuit including classical preprocessing within the QNode."""
+from itertools import product
+from functools import wraps
+from collections import OrderedDict
+from inspect import signature
+import numpy as np
+import pennylane as qml
+from .utils import _get_spectrum, _join_spectra
+
+
+def _process_ids(encoding_args, argnum, qnode):
+    r"""Process the passed ``encoding_args`` and ``argnum`` or infer them from
+    the QNode signature.
+
+    Args:
+        encoding_args (dict[str, list[tuple]] or set): Parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``. If a ``set``, all values are set to ``Ellipsis``
+        argnum (list[int]): Numerical indices for arguments
+        qnode (QNode): QNode to infer the ``encoding_args`` and ``argnum`` from
+            if both are ``None``
+    Returns:
+        OrderedDict[str, list[tuple]]: Ordered parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``
+        list[int]: Numerical indices for arguments
+
+    If ``encoding_args`` are passed, they take precedence over ``argnum``.
+    Passing a set with ``keys`` is an alias for ``{key: ... for key in keys}``.
+    If both, ``encoding_args`` and ``argnum`` are ``None``, all QNode arguments
+    of ``qnode`` that do not have a default value defined are included.
+    Of these arguments, all elements are included in ``encoding_args``
+    """
+    sig_pars = signature(qnode.func).parameters
+    arg_names = list(sig_pars.keys())
+    arg_names_no_def = [name for name, par in sig_pars.items() if par.default is par.empty]
+
+    if encoding_args is None:
+        if argnum is None:
+            encoding_args = OrderedDict((name, ...) for name in arg_names_no_def)
+            argnum = list(range(len(arg_names_no_def)))
```

Comment:
How come both `encoding_args` and `argnum` have to be returned?

Reply:
`encoding_args` for the logic within `advanced_spectrum` and `argnum` for the call of `classical_jacobian` :)